akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -468,8 +465,47 @@ private[stream] class MaterializedValuePublisher extends Publisher[Any] {
 private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Module) {
   import StreamLayout._
 
-  private val subscribers = collection.mutable.HashMap[InPort, Subscriber[Any]]().withDefaultValue(null)
-  private val publishers = collection.mutable.HashMap[OutPort, Publisher[Any]]().withDefaultValue(null)
+  private var subscribersStack: List[mutable.Map[InPort, Subscriber[Any]]] =
+    mutable.Map.empty[InPort, Subscriber[Any]].withDefaultValue(null) :: Nil
+  private var publishersStack: List[mutable.Map[OutPort, Publisher[Any]]] =
+    mutable.Map.empty[OutPort, Publisher[Any]].withDefaultValue(null) :: Nil
+  private var moduleStack: List[Module] = topLevel :: Nil
+
+  private def subscribers: mutable.Map[InPort, Subscriber[Any]] = subscribersStack.head
+  private def publishers: mutable.Map[OutPort, Publisher[Any]] = publishersStack.head
+  private def currentLayout: Module = moduleStack.head
+
+  // Enters a copied module and establishes a scope that prevents internals to leak out and interfere with copies
+  // of the same module.
+  private def enterScope(enclosing: CopiedModule): Unit = {
+    subscribersStack ::= mutable.Map.empty.withDefaultValue(null)
+    publishersStack ::= mutable.Map.empty.withDefaultValue(null)
+    moduleStack ::= enclosing.copyOf
+  }
+
+  // Exits the scope of the copied module and propagates Publishers/Subscribers to the enclosing scope assigning
+  // them to the copied ports instead of the original ones (since there might be multiple copies of the same module
+  // leading to port identity collisions)
+  private def exitScope(enclosing: CopiedModule): Unit = {
```

Comment:
see line 483, we don't save the wrapper on the stack but we save the wrapped module. Btw the only reason for that to be stateful so that I don't need to thread every current state through the subclasses (like ActorMaterializerImpl)